### PR TITLE
[ao][sparsity] Added the Nearly Diagonal Sparsifier

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78448

This sparsifier creates a nearly diagonal mask to be applied to the weight matrix.
    Nearly Diagonal Matrix is a matrix that contains non-zero elements near the diagonal and the rest are zero.
    An example of a nearly diagonal matrix with degree (or nearliness) 3 and 5 are follows respectively.
    1 1 0 0       1 1 1 0
    1 1 1 0       1 1 1 1
    0 1 1 1       1 1 1 1
    0 0 1 1       0 1 1 1
    Note that a nearly diagonal matrix with degree 1 is just a matrix with main diagonal populated

    This sparsifier is controlled by one variable:
    1. `nearliness` defines the number of non-zero diagonal lines that are closest to the main diagonal.
        Currently - supports only odd number

    Note:
        This can be accelerated (vectorized) once the Spdiagonal feature (PR: #78439) is landed or the banded matrix
        feature is landed: https://stackoverflow.com/questions/52463972/generating-banded-matrices-using-numpy

Test Plan:

```
python test/test_ao_sparsity.py TestNearlyDiagonalSparsifier
```

Differential Revision: [D36905916](https://our.internmc.facebook.com/intern/diff/D36905916)

### CODE REVIEW

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        if nearliness % 2 == 0:
+            raise ValueError("nearliness can only be an odd number")
+        dist_to_diagonal = nearliness // 2
+        # check
+        if dist_to_diagonal >= min(height, width):
+            raise ValueError("nearliness cannot be larger than the "
+                            "dimensions of weight matrix.")
+        # create mask
+        for row in range(0, height):
```
Comment (vadimkantorov) [inline torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:48]: just curious, can this be done in a vectorized way? (maybe a good moment to introduce a toeplitz matrix constructor? some options from numpy at https://stackoverflow.com/questions/52463972/generating-banded-matrices-using-numpy)

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        # check
+        if dist_to_diagonal >= min(height, width):
+            raise ValueError("nearliness cannot be larger than the "
+                            "dimensions of weight matrix.")
+
+        for row in range(0, height):
+            # Bounds of entries that needs to be set to 1
+            low = max(0, row - dist_to_diagonal)
+            high = min(width, row + dist_to_diagonal + 1)
+            mask[row, low:high] = torch.ones(high - low, device=mask.device)
```
Comment (vadimkantorov) [inline torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:52]: could it just be `mask[row, low:high].fill_(1)`?

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:48]: Thanks for the suggestion. We could do this by identifying the entries that needs to be set and just fill those entries.  Have implemented this. But not sure how we could use a Toeplitz matrix to accomplish this.

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py (hunk 2 above)
Comment (macandro96) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:52]: That makes sense. I'll amend my commit and raise a pr again.

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:48]: Correct

File: test/ao/sparsity/test_sparsifier.py [hunk 3]
```diff
@@ -346,3 +346,119 @@ def test_sparsity_levels(self):
...
+                sparsifier.squash_mask()
+                model.eval()
+
+                layer = getattr(model, layer_name)
+                # verify that mask created corresponds to the nearliness
+                _verify_nearliness(layer.weight, nearliness)
+
+
+# helper function to verify nearliness of a mask
+def _verify_nearliness(mask: torch.Tensor, nearliness: int):
```
Comment (HDCharles) [inline test/ao/sparsity/test_sparsifier.py:453]: we tend to put these helpers in the test class

**Review by HDCharles — COMMENTED** [review]
should include tags [ao][sparsity] in the PR title

File: test/ao/sparsity/test_sparsifier.py (hunk 3 above)
Comment (macandro96) [reply test/ao/sparsity/test_sparsifier.py:453]: Got it!

File: test/allowlist_for_publicAPI.json [hunk 4]
```diff
@@ -514,6 +514,9 @@
     "Tuple",
     "reduce"
   ],
+  "torch.ao.sparsity.sparsifier.nearly_diagonal_sparsifier": [
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:517]: You should not add new entries to this file.
Make sure to follow the guidelines from the error message to make this properly public or private depending on what it should be.

**Review by albanD — CHANGES_REQUESTED** [review]

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py (hunk 1 above)
Comment (vadimkantorov) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:47]: So a toeplitz matrix or banded matrix constructor should be helpful for this

Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:47]: @vadimkantorov Because there is no toeplitz constructor kernel implemented, do you think it's fine to put it as a comment in the docstring, and leave it as is?

File: torch/ao/sparsity/__init__.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 # Sparsifier
 from .sparsifier.base_sparsifier import BaseSparsifier
 from .sparsifier.weight_norm_sparsifier import WeightNormSparsifier
+from .sparsifier.nearly_diagonal_sparsifier import NearlyDiagonalSparsifier
```
Comment (z-a-f) [inline torch/ao/sparsity/__init__.py:8]: Add the following in this file:

```python
__all__ = [
    'NearlyDiagonalSparsifier',
]
```

I think that should fix the issue. I will all other imports in there in the later PRs. cc @albanD

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py (hunk 1 above)
Comment (vadimkantorov) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:47]: yep, I think so! then if it's implemented, this code can be updated

File: torch/ao/sparsity/__init__.py (hunk 5 above)
Comment (z-a-f) [reply torch/ao/sparsity/__init__.py:8]: Actually scratch that -- We probs need to avoid exposing the `BaseSparsifier` in the `nearly_diagonal...` file

**Review by z-a-f — APPROVED** [review]
LGTM

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:51]: Added the comment in doc-string. Will update once implemented.

File: torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    def update_mask(self, layer, nearliness,
+                    **kwargs):
+        mask = layer.parametrizations.weight[0].mask
+        mask.data = torch.zeros_like(mask)
+        if nearliness <= 0:
+            return
+
+        weight = layer.weight
+        height, width = weight.shape
```
Comment (z-a-f) [inline torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:42]: nit:
```
height, width = layer.weight.shape
```

**Comment by z-a-f** [comment]
FYI, I checked different ways of implementing the logic, and it looks like vectorization might give a slight edge: https://colab.research.google.com/drive/1USyM4mTuAFml5ntNX-TuUPVtY7AJDtAB?usp=sharing

I won't block it on this, just an fyi :)

**Review by HDCharles — APPROVED** [review]
looks good though there's a failing CI test that you need to resolve before landing, not sure what here could be causing it though, probably just rebase and see if it goes away.

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36905916).

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36905916).

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36905916).
